### PR TITLE
feat: trivy scan and kusto export in the same script

### PR DESCRIPTION
**Pull request by jason1028kr**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
To integrate Kusto module into our vhd pipeline
- call vunl-to-kusto module within the trivy-scan script.
- move storage account upload logic into the `trivy-scan.sh.`
- Keep the current image table
- 
**Which issue(s) this PR fixes**:

Fixes #

**Requirements**:

- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Special notes for your reviewer**:

**Release note**:

```
none
```

### CODE REVIEW

File: vhdbuilder/packer/trivy-scan.sh [hunk 1]
```diff
@@ -1,41 +1,142 @@
...
+    ./trivy --scanners vuln image --ignore-unfixed --severity HIGH,CRITICAL -f table ${CONTAINER_IMAGE} >> ${TRIVY_REPORT_IMAGE_TABLE_PATH} || true
 
-for image in $IMAGE_LIST; do
-    ./trivy --scanners vuln image --ignore-unfixed --severity HIGH,CRITICAL -f table $image >> ${TRIVY_REPORT_TABLE_PATH} || true
+    # export to Kusto, one by one
+    BASE_CONTAINER_IMAGE=$(basename ${CONTAINER_IMAGE})
+    TRIVY_REPORT_IMAGE_JSON_PATH=${TRIVY_REPORT_DIRNAME}/trivy-report-image-${BASE_CONTAINER_IMAGE}.json
+    ./trivy --scanners vuln image -f json --ignore-unfixed --severity HIGH,CRITICAL -o ${TRIVY_REPORT_IMAGE_JSON_PATH} $CONTAINER_IMAGE || true
+
+    if [[ -f ${TRIVY_REPORT_IMAGE_JSON_PATH} ]]; then
```
Comment (djsly) [inline vhdbuilder/packer/trivy-scan.sh:112]: why do we need a `if` statement here and not for the `rootfs` case. WE don't have  a json file created with there isn't any vulnerabilities ?

File: vhdbuilder/packer/trivy-scan.sh [hunk 2]
```diff
@@ -1,41 +1,142 @@
...
+for CONTAINER_IMAGE in $IMAGE_LIST; do
+    # append to table
+    ./trivy --scanners vuln image --ignore-unfixed --severity HIGH,CRITICAL -f table ${CONTAINER_IMAGE} >> ${TRIVY_REPORT_IMAGE_TABLE_PATH} || true
 
-for image in $IMAGE_LIST; do
-    ./trivy --scanners vuln image --ignore-unfixed --severity HIGH,CRITICAL -f table $image >> ${TRIVY_REPORT_TABLE_PATH} || true
+    # export to Kusto, one by one
+    BASE_CONTAINER_IMAGE=$(basename ${CONTAINER_IMAGE})
+    TRIVY_REPORT_IMAGE_JSON_PATH=${TRIVY_REPORT_DIRNAME}/trivy-report-image-${BASE_CONTAINER_IMAGE}.json
+    ./trivy --scanners vuln image -f json --ignore-unfixed --severity HIGH,CRITICAL -o ${TRIVY_REPORT_IMAGE_JSON_PATH} $CONTAINER_IMAGE || true
```
Comment (djsly) [inline vhdbuilder/packer/trivy-scan.sh:110]: why are we `|| true` , ignoring failure with the scanner ?

File: vhdbuilder/packer/trivy-scan.sh [hunk 3]
```diff
@@ -1,41 +1,142 @@
...
+
+        rm ${TRIVY_REPORT_IMAGE_JSON_PATH} || true
+    fi
 done
 
 rm ./trivy 
 
-chmod a+r "${TRIVY_REPORT_JSON_PATH}"
-chmod a+r "${TRIVY_REPORT_TABLE_PATH}"
+chmod a+r "${TRIVY_REPORT_ROOTFS_JSON_PATH}"
```
Comment (djsly) [inline vhdbuilder/packer/trivy-scan.sh:129]: ain't you removing the report on line `123` ?

File: vhdbuilder/packer/trivy-scan.sh [hunk 4]
```diff
@@ -1,41 +1,142 @@
...
+chmod a+r "${TRIVY_REPORT_IMAGE_TABLE_PATH}"
+
+az storage blob upload --file ${TRIVY_REPORT_ROOTFS_JSON_PATH} \
```
Comment (djsly) [inline vhdbuilder/packer/trivy-scan.sh:132]: why are we uploading here and in the vid-scanning file ?

**Review by djsly — CHANGES_REQUESTED** [review]

File: vhdbuilder/packer/trivy-scan.sh [hunk 5]
```diff
@@ -1,41 +1,142 @@
...
+mkdir -p "$(dirname "${TRIVY_REPORT_DIRNAME}")"
 
 wget "https://github.com/aquasecurity/trivy/releases/download/v${TRIVY_VERSION}/trivy_${TRIVY_VERSION}_${TRIVY_ARCH}.tar.gz"
 tar -xvzf "trivy_${TRIVY_VERSION}_${TRIVY_ARCH}.tar.gz"
 rm "trivy_${TRIVY_VERSION}_${TRIVY_ARCH}.tar.gz"
 chmod a+x trivy 
 
-./trivy --scanners vuln rootfs -f json --skip-dirs /var/lib/containerd --ignore-unfixed --severity HIGH,CRITICAL -o "${TRIVY_REPORT_JSON_PATH}" /
+# pull vuln-to-kusto binary
+az storage blob download --auth-mode login --account-name vulntokustobinary -c vuln-to-kusto \
```
[the remark this answers is not in the document]
Comment (jason1028kr) [reply vhdbuilder/packer/trivy-scan.sh:79]: yes

File: vhdbuilder/packer/trivy-scan.sh [hunk 6]
```diff
@@ -1,41 +1,142 @@
...
+    sudo apt update
+    sudo apt install -y python3-pip
+    pip install azure-cli
+    export PATH="/home/$TEST_VM_ADMIN_USERNAME/.local/bin:$PATH"
+    CHECKAZ=$(pip freeze | grep "azure-cli==")
+    if [[ -z $CHECKAZ ]]; then
+        echo "Azure CLI is not installed properly."
+        exit 1
+    fi
+elif [[ "$OS_SKU" == "Ubuntu" ]] && [[ "$OS_VERSION" == "22.04" ]] && [[ "${ARCHITECTURE,,}" != "arm64" ]]; then
```
Comment (cameronmeissner) [inline vhdbuilder/packer/trivy-scan.sh:38]: I know @AlisonB319 is working on onboarding 2404 to scanning, might want to sync with her on this

File: vhdbuilder/packer/trivy-scan.sh [hunk 7]
```diff
@@ -1,41 +1,142 @@
...
+# pull vuln-to-kusto binary
+az storage blob download --auth-mode login --account-name vulntokustobinary -c vuln-to-kusto \
+    --blob-url https://vulntokustobinary.blob.core.windows.net/vuln-to-kusto/v0.0.1-5a0563d8080/vuln-to-kusto-vhd_linux_${GO_ARCH} \
+    --file ./vuln-to-kusto-vhd
+chmod a+x vuln-to-kusto-vhd
+
+# shellcheck disable=SC2155
+export PATH="$(pwd):$PATH"
+
+./trivy --scanners vuln rootfs -f json --skip-dirs /var/lib/containerd --ignore-unfixed --severity HIGH,CRITICAL -o "${TRIVY_REPORT_ROOTFS_JSON_PATH}" /
```
[the remark this answers is not in the document]
Comment (jason1028kr) [reply vhdbuilder/packer/trivy-scan.sh:87]: also added `severity` to variable group

File: vhdbuilder/packer/trivy-scan.sh (hunk 2 above)
Comment (jason1028kr) [reply vhdbuilder/packer/trivy-scan.sh:110]: I believe this was for skipping over images that trivy fails to scan
"
./vuln-to-kusto-vhd scan-report --vhd-vhdname=2204-gen2-containerd --vhd-ossku=Ubuntu --vhd-osversion=22.04 --vhd-skuname=2204gen2containerd --kusto-endpoint=https://aksdevinfra.westus2.kusto.windows.net/ --kusto-database=AKSShieldGuard --kusto-table=AKSContainerCVEScanner /opt/azure/containers/trivy-report-image-aks-node-ca-watcher:master.240530.2.json\n2024/08/08 01:56:13 WARN no environment resolved, fallback to unknown environment component=environment-detector\n2024/08/08 01:56:13 WARN no build source detected, fallback to unknown build source component=build-source-detector\n+ rm /opt/azure/containers/trivy-report-image-aks-node-ca-watcher:master.240530.2.json\n+ for CONTAINER_IMAGE in $IMAGE_LIST\n+ ./trivy --scanners vuln image --ignore-unfixed --severity HIGH,CRITICAL -f table mcr.microsoft.com/aks/aks-node-ca-watcher:static\n2024-08-08T01:56:14.311Z\t\u001b[34mINFO\u001b[0m\tVulnerability scanning is enabled\n2024-08-08T01:56:14.477Z\t\u001b[31mFATAL\u001b[0m\timage scan error: scan error: unable to initialize a scanner: unable to initialize a docker scanner: 4 errors occurred:\n\t* unable to inspect the image (mcr.microsoft.com/aks/aks-node-ca-watcher:static): Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?\n\t* unable to initialize Podman client: no podman socket found: stat podman/podman.sock: no such file or directory\n\t* image not found in containerd store: mcr.microsoft.com/aks/aks-node-ca-watcher:static\n\t* GET https://mcr.microsoft.com/v2/aks/aks-node-ca-watcher/manifests/static: MANIFEST_UNKNOWN: manifest tagged by \"static\" is not found;
"

File: vhdbuilder/packer/trivy-scan.sh (hunk 1 above)
Comment (jason1028kr) [reply vhdbuilder/packer/trivy-scan.sh:112]: Right, adding if block to rootfs also

File: vhdbuilder/packer/trivy-scan.sh (hunk 6 above)
Comment (jason1028kr) [reply vhdbuilder/packer/trivy-scan.sh:38]: yep, thanks for the reminder

File: vhdbuilder/packer/trivy-scan.sh (hunk 3 above)
Comment (jason1028kr) [reply vhdbuilder/packer/trivy-scan.sh:129]: So we 
- remove  `trivy json` output for images after uploading
- keep `trivy rootfs` output + `trivy table` output, for uploading to storage account and leaving them as pipeline artifacts.

File: vhdbuilder/packer/trivy-scan.sh (hunk 4 above)
Comment (jason1028kr) [reply vhdbuilder/packer/trivy-scan.sh:132]: I've removed `vhdbuilder/packer/vhd-scanning-exe-on-vm.sh` and moved uploading logic to here. Within `vhd-scanning.sh,` we simply download them for pipeline artifacts.

**Review by djsly — APPROVED** [review]

**Pull request merged**